src/main/java/org/mastodon/graph/branch/BranchGraph.java

```diff
@@ -0,0 +1,230 @@
+ *
+ * <pre>
+ *     v1 &#8594; v2 &#8594; v3
+ *   /             \
+ * v0               v7
+ *   \             /
+ *     v4 &#8594; v5 &#8594; v6
+ * </pre>
+ *
+ * In the branch graph, this becomes:
+ *
+ * <pre>
+ *     ____
+ *    /    \
+ * bv0      bv1
+ *    \____/
+ * </pre>
+ *
+ * The branch graph can also handle ring-link structures in the linked graph. In
+ * that case, such a ring;
+ *
+ * <pre>
+ *     v1 &#8594; v2 &#8594; v3
+ *   /             \
+ * v0               v4
+ *   \             /
+ *     v7 &#8592; v6 &#8592; v5
+ * </pre>
+ *
+ * is represented by a single vertex having a loop-edge:
+ *
+ * <pre>
+ *    _
+ * bv0 \
+ *  \__/
+ * </pre>
+ *
+ * In such a loop, the vertex linked to <code>bv0</code> and the edge linked to
+ * the branch edge are determined by the order in which the vertices and edges
+ * are added in the linked graph.
+ * <p>
+ * The branch graph is defined and tested only for <b>simple, directed
+ * graphs</b> as linked graph. Using any other classes of graphs will result in
+ * unexpected behavior.
+ *
+ * @param <BV>
+ *            the type of the branch vertices.
+ * @param <BE>
+ *            the type of the branch edges.
+ * @param <V>
+ *            the type of linked vertices.
+ * @param <E>
+ *            the type of linked edges.
+ */
+public interface BranchGraph<
+	BV extends Vertex< BE >,
+	BE extends Edge< BV >,
+	V extends Vertex< E >,
+	E extends Edge< V > >
+		extends ListenableReadOnlyGraph< BV, BE >
+{
+
+	/**
+	 * Returns the edge linked to the specified branch edge. The linked edge is
+	 * the single outgoing edge of the branch extremity.
+	 *
+	 * @param be
+	 *            the branch edge.
+	 * @param ref
+	 *            a reference to a linked graph edge used for retrieval.
+	 *            Depending on concrete implementation of the linked graph, this
+	 *            object can be cleared, ignored or re-used.
+	 * @return the linked edge.
+	 */
+	public E getLinkedEdge( BE be, E ref );
+
+	/**
+	 * Returns the vertex linked to the specified branch vertex. The linked
+	 * vertex is a branch extremity.
+	 *
+	 * @param bv
+	 *            the branch vertex.
+	 * @param ref
+	 *            a reference to a linked graph vertex used for retrieval.
+	 *            Depending on concrete implementation of the linked graph, this
+	 *            object can be cleared, ignored or re-used.
+	 * @return the linked vertex.
+	 */
+	public V getLinkedVertex( BV bv, V ref );
+
+	/**
+	 * Returns the branch edge linked to the specified edge.
+	 *
+	 * @param edge
+	 *            the linked edge.
+	 * @param ref
+	 *            a reference object to the branch edge used for retrieval.
+	 * @return a branch edge.
+	 */
+	public BE getBranchEdge( E edge, BE ref );
+
+	/**
+	 * Returns the branch vertex linked to the specified edge if it belongs to a
+	 * branch. Returns <code>null</code> if the specified vertex is a branch
+	 * extremity.
```

Comment:
This seems to confuse edge/vertex. "Returns the branch vertex", but returns a branch edge. "...to the specified edge", but takes a vertex argument. Please revise.